go/test/endtoend/vtgate/tabletbalancer/balancer_test.go

```diff
@@ -0,0 +1,483 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tabletbalancer
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/test/endtoend/cluster"
+	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
+)
+
+// Helper functions for e2e testing of balancer modes
+
+// executeReplicaQueries executes queries against replicas and verifies they succeed
+func executeReplicaQueries(t *testing.T, conn *mysql.Conn, numQueries int) map[int64]int {
+	t.Helper()
+
+	// Use @replica to target replica tablets
+	_, err := conn.ExecuteFetch("USE @replica", 1, false)
+	require.NoError(t, err)
+
+	counts := make(map[int64]int)
+
+	// Execute queries and verify they succeed
+	for range numQueries {
+		res, err := conn.ExecuteFetch("SELECT @@server_id, id FROM balancer_test LIMIT 1", 10, false)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(res.Rows), "expected one row from balancer_test")
+
+		serverID, err := res.Rows[0][0].ToInt64()
+		require.NoError(t, err)
+		counts[serverID]++
+	}
+
+	return counts
+}
+
+func getTabletIDFromServerID(t *testing.T, tablets []*cluster.Vttablet) map[string]int64 {
+	aliases := make(map[string]int64)
+
+	for _, tablet := range tablets {
+		id, err := tablet.VttabletProcess.QueryTablet("SELECT @@server_id", tablet.VttabletProcess.Keyspace, false)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(id.Rows), "expected one row for server_id query")
+
+		serverID, err := id.Rows[0][0].ToInt64()
+		assert.NoError(t, err)
+		aliases[tablet.Alias] = serverID
+	}
+
+	assert.Equal(t, len(aliases), 6, "expected six tablet aliases, got: %d", len(aliases))
+
+	return aliases
+}
+
+// TestCellModeBalancer tests the default "cell" mode which shuffles tablets in the local cell
+func TestCellModeBalancer(t *testing.T) {
+	// Start vtgate in cell1 with cell mode (default)
+	vtgateProcess := cluster.VtgateProcessInstance(
+		clusterInstance.GetAndReservePort(),
+		clusterInstance.GetAndReservePort(),
+		clusterInstance.GetAndReservePort(),
+		cell1,
+		fmt.Sprintf("%s,%s", cell1, cell2), // watch both cells but should prefer local
+		clusterInstance.Hostname,
+		"replica",
+		clusterInstance.TopoProcess.Port,
+		clusterInstance.TmpDirectory,
+		[]string{
+			"--vtgate-balancer-mode", "cell",
+		},
+		plancontext.PlannerVersion(0),
+	)
+	require.NoError(t, vtgateProcess.Setup())
+	defer vtgateProcess.TearDown()
+
+	// Verify vtgate started successfully
+	require.True(t, vtgateProcess.WaitForStatus())
+
+	// Wait for tablets to be discovered
+	time.Sleep(2 * time.Second)
```

Comment:
I would recommend waiting for some condition using require.Eventually if possible. We've found these kinds of sleeps to inevitably lead to flaky tests as the performance characteristics of GitHub runners are unpredictable (I've seen 1+ second CPU stalls).